cmd/kritis/signer/server.go

```diff
@@ -0,0 +1,228 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"google.golang.org/genproto/googleapis/devtools/containeranalysis/v1beta1/discovery"
+	grafeaspb "google.golang.org/genproto/googleapis/devtools/containeranalysis/v1beta1/grafeas"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+)
+
+type EvaluationStatus string
+
+const (
+	Ok        EvaluationStatus = "ok"
+	Signed    EvaluationStatus = "signed"
+	NotSigned EvaluationStatus = "not-signed"
+	Failed    EvaluationStatus = "failed"
+)
+
+type SignRequest struct {
+	Image string
+}
+
+type SignResponse struct {
+	Image      string           `json:"image,omitempty"`
+	Status     EvaluationStatus `json:"status,omitempty"`
+	Message    string           `json:"message,omitempty"`
+	Violations []string         `json:"violations,omitempty"`
+}
+
+// write the sign `response` to `w` with status `code`
+func WriteResponse(w http.ResponseWriter, response SignResponse, code int) {
+	body, err := json.Marshal(response)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	glog.Infof("%s\t%s", response.Image, response.Message)
+	w.Header().Set("Content-Type", "application/json; charset=utf-8")
+	w.WriteHeader(code)
+	w.Write(body)
+}
+
+// Google Pub/Sub message
+type PubSubMessage struct {
+	Message struct {
+		Data []byte `json:"data,omitempty"`
+		ID   string `json:"id"`
+	} `json:"message"`
+	Subscription string `json:"subscription"`
+}
+
+// Google Container Analysis occurrence notification
+type ContainerAnalysisOccurrenceEvent struct {
+	Name             string    `json:"name"`
+	Kind             string    `json:"kind"`
+	NotificationTime time.Time `json:"notificationTime"`
+}
+
+// retrieve the container analysis occurrence associated with this notification
+func GetOccurrence(event ContainerAnalysisOccurrenceEvent) (*grafeaspb.Occurrence, error) {
+	request := grafeaspb.GetOccurrenceRequest{Name: event.Name}
+	response, err := grafeas.GetOccurrence(context.Background(), &request)
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve discovery occurrence %s", event.Name)
+	}
+	return response, nil
+}
+
+// implements `/event` handles a container analysis occurrence notification
+func containerAnalysisEvent(w http.ResponseWriter, r *http.Request) {
+	var m PubSubMessage
+	var event ContainerAnalysisOccurrenceEvent
+	err := json.NewDecoder(r.Body).Decode(&m)
+	if err != nil {
+		http.Error(w, "could not unmarshal the pubsub message", http.StatusBadRequest)
+		return
+	}
+
+	err = json.Unmarshal(m.Message.Data, &event)
+	if err != nil {
+		http.Error(w, "could not unmarshal container analysis occurrence event from pubsub message", http.StatusBadRequest)
+		return
+	}
+
+	if event.Kind != "DISCOVERY" {
+		http.Error(w, "ignoring non-DISCOVERY occurrence", http.StatusOK)
```

Comment:
Perhaps downgrade logging level.  This doesn't seem like some exceptional circumstance: AIUI pubsub will send all the Occurrences, not just the DISCOVERY ones.

Reply:
I did not know that http.Error() prints to stderr too. Fixed it so that all conditions will return an appropriate SignResponse.